kubemarine/packages.py

```diff
@@ -12,64 +12,196 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from copy import deepcopy
 from typing import List, Dict
 
-from kubemarine import yum, system, apt
+from kubemarine import yum, apt
+from kubemarine.core.cluster import KubernetesCluster
 from kubemarine.core.executor import RemoteExecutor
 from kubemarine.core.group import NodeGroup, NodeGroupResult
+from kubemarine.core.yaml_merger import default_merger
 
 
-def enrich_inventory_associations(inventory, cluster):
-    os_family = system.get_os_family(cluster)
+ERROR_GLOBAL_ASSOCIATIONS_REDEFINED_MULTIPLE_OS = \
+    "It is not supported to customize services.packages.associations section " \
+    "if nodes have different OS families. " \
+    "Please move the section to corresponding services.packages.associations.<os_family> section."
 
-    associations = inventory['services']['packages']['associations']
-    if not associations.get(os_family):
-        # already enriched
+ERROR_MULTIPLE_PACKAGE_VERSIONS_DETECTED = \
+    "Multiple package versions detected %s for package '%s'. " \
+    "Align them to the single version manually or using corresponding task of install procedure. " \
+    "Alternatively, specify cache_versions=false for corresponding association."
+
+
+def enrich_inventory_associations(inventory, cluster: KubernetesCluster):
+    associations: dict = inventory['services']['packages']['associations']
+    os_propagated_associations = {}
+
+    # move associations for OS families as-is
+    for association_name in get_associations_os_family_keys():
+        os_propagated_associations[association_name] = associations.pop(association_name)
+
+    inventory['services']['packages']['associations'] = os_propagated_associations
+
+    # Check remained associations section if they are customized at global level.
+    if associations:
+        os_family = cluster.get_os_family()
+        if os_family == 'multiple':
+            raise Exception(ERROR_GLOBAL_ASSOCIATIONS_REDEFINED_MULTIPLE_OS)
```

Comment:
New restriction. When migrating of OS, it is necessary to move all customizations to OS family specific sections.